projects/experimental/directives/ripple/ripple.directive.ts

```diff
@@ -0,0 +1,121 @@
+import {Directive, HostListener, Inject, Input, Renderer2, Self} from '@angular/core';
+import {
+    ALWAYS_FALSE_HANDLER,
+    ALWAYS_TRUE_HANDLER,
+    TuiDestroyService,
+    TuiDirectiveStylesService,
+    tuiPx,
+    tuiTypedFromEvent,
+} from '@taiga-ui/cdk';
+import {merge, Observable, race, timer} from 'rxjs';
+import {
+    defaultIfEmpty,
+    filter,
+    map,
+    switchMap,
+    take,
+    takeUntil,
+    tap,
+} from 'rxjs/operators';
+
+import {TuiRippleStylesComponent} from './ripple-styles.component';
+
+const TOUCH_MOVE_DELAY = 100;
+
+const RIPPLE_ON = 'tuiRippleOn';
+
+const RIPPLE_OFF = 'tuiRippleOff';
+
+function tuiAnimationEnd(ripple: HTMLElement, animationName: string): Observable<void> {
+    return tuiTypedFromEvent(ripple, 'animationend')
+        .pipe(filter(event => animationName === event.animationName))
+        .pipe(map(() => undefined));
+}
+
+@Directive({
+    selector: '[tuiRipple]',
+    providers: [TuiDestroyService],
+})
+export class TuiRippleDirective {
+    @Input('tuiRipple')
+    selector = 'button';
+
+    constructor(
+        @Inject(TuiDirectiveStylesService) directiveStyles: TuiDirectiveStylesService,
+        @Inject(Renderer2) private readonly renderer: Renderer2,
+        @Self() @Inject(TuiDestroyService) private readonly destroy$: TuiDestroyService,
+    ) {
+        directiveStyles.addComponent(TuiRippleStylesComponent);
+    }
+
+    @HostListener('touchstart', [
+        '$event.touches[0].clientX',
+        '$event.touches[0].clientY',
+        '$event.target',
+    ])
+    @HostListener('mousedown', ['$event.clientX', '$event.clientY', '$event.target'])
+    start(clientX: number, clientY: number, target: HTMLElement): void {
+        const nativeElement = target.closest(this.selector);
+
+        if (!(nativeElement instanceof HTMLElement)) {
+            return;
+        }
+
+        const {width, height, top, left} = nativeElement.getBoundingClientRect();
+
+        const radius = Math.sqrt(width * width + height * height);
+        const dimension = radius * 2;
+        const x = clientX - left - radius;
+        const y = clientY - top - radius;
+        const ripple = this.renderer.createElement('div');
```

Comment:
I think it would be better to move creating this element to a separate method for readability.

Reply:
fixed